### PR TITLE
Feature/get system token

### PR DESCRIPTION
**Pull request by barbararochazup**
## Issue Description

Implementation of Get system token endpoint

### Request (if appropriate):
**URL:** GET /api/v1/system-token/{id}

Payload:
```json
{
  "id": "06666a99-7800-4c83-af2c-a1a1d5e1bb2e",
  "name": "Token Test",
  "permissions": [
    {
      "id": "12666a99-7800-4c83-af2c-a1a1d5e1bb3f",
      "name": "MAINTAINER"
    }
  ],
  "cretedAt": "2020-02-26 13:00:00",
  "author": "charlescd@email.com"
}
```

> Status: **404 NOT FOUND**

Payload:
```json
{
    "id": "5832ec91-88c9-4ca5-861e-af49e655048e",
    "message": "Not found token",
    "Type": "NotFoundError",
    "timestamp": "1614345707",
    "meta": null
}
```

> Status: **500 INTERNAL ERROR**

Payload:
```json
{
    "id": "5832ec91-88c9-4ca5-861e-af49e655048e",
    "message": "Internal error details",
    "Type": "InternaldError",
    "timestamp": "1614345707",
    "meta": null
}
```

> Status: **400 BAD REQUEST**

Payload:
```json
{
    "id": "9cebaea5-69e6-4865-b0ec-aefa0ceadc65",
    "message": "Parse id failed",
    "Type": "ParseError",
    "timestamp": "1614356891",
    "meta": null
}
```

### CODE REVIEW

**Review by pedroborgeszup — APPROVED** [review]

File: gate/internal/repository/system_token_repository.go [hunk 1]
```diff
@@ -1,3 +1,59 @@
...
+	return systemToken, nil
+}
+
+func (systemTokenRepository systemTokenRepository) FindById(id uuid.UUID) (domain.SystemToken, error)  {
+	var systemToken models.SystemToken
+
+	res := systemTokenRepository.db.Model(models.SystemToken{}).Where("id = ?", id).First(&systemToken)
+	if res.Error != nil {
+		if res.Error.Error() == "record not found" {
+			return handlerError("Not found token", "unit.GetById.First", res.Error, logging.NotFoundError)
```
[the remark this answers is not in the document]
Comment (barbararochazup) [reply gate/internal/repository/system_token_repository.go]: Done

File: gate/web/api/handlers/system_token_handler.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+func GetSystemToken(getSystemToken systemTokenInteractor.GetSystemToken) echo.HandlerFunc {
+	return func(echoCtx echo.Context) error {
+		ctx := echoCtx.Request().Context()
+		uuid, parseErr := uuidPkg.Parse(echoCtx.Param("id"))
+		if parseErr != nil {
+			logging.LogErrorFromCtx(ctx, parseErr)
+			return echoCtx.JSON(http.StatusBadRequest, logging.NewError("Parse id failed", parseErr, logging.ParseError, nil))
+		}
+
+		user, err := getSystemToken.Execute(uuid)
```
Comment (pedroborgeszup) [inline gate/web/api/handlers/system_token_handler.go:21]: Shouldn't the getSystemToken result variable be `systemToken` instead of `user`?

**Review by pedroborgeszup — CHANGES_REQUESTED** [review]

**Review by pedroborgeszup — APPROVED** [review]

File: gate/web/api/handlers/system_token_handler.go (hunk 2 above)
Comment (barbararochazup) [reply gate/web/api/handlers/system_token_handler.go]: Done

File: gate/tests/unit/system_token_test.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	require.NotNil(st.T(), a)
+	require.Nil(st.T(), err)
+}
+
+func (st *SystemTokenSuite) TestErrorGetByIDNotFound() {
+	st.systemTokenRepository.On("FindById", mock.Anything).Return(domain.SystemToken{}, logging.NewError("Not found error", logging.CustomError{} ,  logging.NotFoundError, nil))
+	_, err := st.getSystemToken.Execute(uuid.New())
+	require.Error(st.T(), err)
+	require.Equal(st.T(), logging.NotFoundError, logging.GetErrorType(err))
+
```
Comment (monicaribeiro) [inline gate/tests/unit/system_token_test.go:46]: Remove extra lines.

**Review by monicaribeiro — CHANGES_REQUESTED** [review]

File: gate/tests/unit/system_token_test.go (hunk 3 above)
Comment (barbararochazup) [reply gate/tests/unit/system_token_test.go]: Done

File: gate/tests/unit/system_token_test.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+func (st *SystemTokenSuite) SetupTest() {
+	st.SetupSuite()
+}
+
+func TestSuite(t *testing.T) {
+	suite.Run(t, new(SystemTokenSuite))
+}
+
+func (st *SystemTokenSuite) TestGetByID() {
+	st.systemTokenRepository.On("FindById", mock.Anything).Return(domain.SystemToken{ID: uuid.New()}, nil)
```
Comment (monicaribeiro) [inline gate/tests/unit/system_token_test.go:35]: Suggestion:

To make sure that the correct input is going to be received, you could initialize a system token and reference it here:
```suggestion
	st.systemTokenRepository.On("FindById", systemToken.ID).Return(systemToken, nil)
```

File: gate/tests/unit/system_token_test.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	a, err := st.getSystemToken.Execute(uuid.New())
+	require.NotNil(st.T(), a)
+	require.Nil(st.T(), err)
```
Comment (monicaribeiro) [inline gate/tests/unit/system_token_test.go:38]: You could add another verification to make sure that the correct id was passed to the FindById method. Also, check if FindById was called once.

File: gate/tests/unit/system_token_test.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+func (st *SystemTokenSuite) TestErrorGetByIDNotFound() {
+	st.systemTokenRepository.On("FindById", mock.Anything).Return(domain.SystemToken{}, logging.NewError("Not found error", logging.CustomError{} ,  logging.NotFoundError, nil))
```
Comment (monicaribeiro) [inline gate/tests/unit/system_token_test.go:42]: Tha same comments I made before are applicable here.

Comment (monicaribeiro) [reply gate/tests/unit/system_token_test.go:42]: the**

File: gate/tests/unit/system_token_test.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	_, err := st.getSystemToken.Execute(uuid.New())
+	require.Error(st.T(), err)
+	require.Equal(st.T(), logging.NotFoundError, logging.GetErrorType(err))
+}
+
+func (st *SystemTokenSuite) TestErrorGetByIDInternalError() {
+	st.systemTokenRepository.On("FindById", mock.Anything).Return(domain.SystemToken{}, logging.NewError("Internal error", logging.CustomError{}, logging.InternalError, nil))
```
Comment (monicaribeiro) [inline gate/tests/unit/system_token_test.go:49]: The same comments I made before are applicable here.

**Review by monicaribeiro — CHANGES_REQUESTED** [review]

File: gate/tests/unit/system_token_test.go (hunk 7 above)
Comment (barbararochazup) [reply gate/tests/unit/system_token_test.go:49]: Done

File: gate/tests/unit/system_token_test.go (hunk 6 above)
Comment (barbararochazup) [reply gate/tests/unit/system_token_test.go:42]: Done

File: gate/tests/unit/system_token_test.go (hunk 5 above)
Comment (barbararochazup) [reply gate/tests/unit/system_token_test.go:38]: Done

**Review by pedroborgeszup — CHANGES_REQUESTED** [review]

File: gate/tests/unit/system_token_test.go [hunk 8]
```diff
@@ -32,22 +33,50 @@ func TestSuite(t *testing.T) {
...
-	_, err := st.getSystemToken.Execute(uuid.New())
+	systemToken := getDummySystemToken()
+
+	st.systemTokenRepository.On("FindById", systemToken.ID).
+		Return(systemToken, logging.NewError("Internal error", logging.CustomError{}, logging.InternalError, nil))
+	result, err := st.getSystemToken.Execute(systemToken.ID)
+
 	require.Error(st.T(), err)
 	require.Equal(st.T(), logging.InternalError, logging.GetErrorType(err))
+	require.Equal(st.T(), result.ID, uuid.Nil)
```
Comment (pedroborgeszup) [inline gate/tests/unit/system_token_test.go]: in an actual call, if an error occurs, the result object is returned empty, different from what is happening in the test

**Review by pedroborgeszup — CHANGES_REQUESTED** [review]

File: gate/tests/unit/system_token/system_token_test.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+func (st *SystemTokenSuite) TestErrorGetByIDNotFound() {
+	systemToken := getDummySystemToken()
+
+	st.systemTokenRepository.On("FindById", systemToken.ID).
+		Return(systemToken, logging.NewError("Not found error", logging.CustomError{} ,  logging.NotFoundError, nil))
+	result, err := st.getSystemToken.Execute(systemToken.ID)
+
+	require.Error(st.T(), err)
+	require.Equal(st.T(), logging.NotFoundError, logging.GetErrorType(err))
+	require.Equal(st.T(), result.ID, uuid.Nil)
```
Comment (pedroborgeszup) [inline gate/tests/unit/system_token/system_token_test.go:56]: in an actual call, if an error occurs, the result object is returned empty, different from what is happening in the test

File: gate/tests/unit/system_token/system_token_test.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+func (st *SystemTokenSuite) TestErrorGetByIDInternalError() {
+	systemToken := getDummySystemToken()
+
+	st.systemTokenRepository.On("FindById", systemToken.ID).
+		Return(systemToken, logging.NewError("Internal error", logging.CustomError{}, logging.InternalError, nil))
+	result, err := st.getSystemToken.Execute(systemToken.ID)
+
+	require.Error(st.T(), err)
+	require.Equal(st.T(), logging.InternalError, logging.GetErrorType(err))
+	require.Equal(st.T(), result.ID, uuid.Nil)
```
Comment (pedroborgeszup) [inline gate/tests/unit/system_token/system_token_test.go:70]: in an actual call, if an error occurs, the result object is returned empty, different from what is happening in the test

**Review by pedroborgeszup — CHANGES_REQUESTED** [review]

**Review by monicaribeiro — APPROVED** [review]

File: gate/web/api/handlers/representation/system_token.go [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+	Author string `json:"author" validate:"required,notblank"`
+}
+
+type SystemTokenResponse struct {
+	ID				uuid.UUID `json:"id"`
+	Name 			string `json:"name"`
+	Permissions 	[]PermissionResponse `json:"permissions"`
+	CreatedAt 		*time.Time `json:"created_at"`
+	RevokedAt  		*time.Time `json:"revoked_at,omitempty"`
+	LastUsedAt 		*time.Time `json:"last_used_atm,omitempty"`
```
[the remark this answers is not in the document]
Comment (barbararochazup) [reply gate/web/api/handlers/representation/system_token.go]: done

**Review by pedroborgeszup — APPROVED** [review]

**Pull request merged**